src/S3/StreamWrapper.php

```diff
@@ -428,7 +430,7 @@ public function dir_readdir()
 
         // Cache the object data for quick url_stat lookups used with
         // RecursiveDirectoryIterator.
-        self::$nextStat = [$key => $stat];
+        self::$nextStat[$key] = $stat;
```

Comment:
I changed nextStat to a hash so that it caches more than just a single element, but rather caches each element as it is iterated. This significantly reduces the number of requests made in the stream wrapper, and these cached items are cleared out when a dir resource is closed.